File: src/main/java/io/pageobject/generator/locator/RepeaterLocatorPartGenerator.java
package io.pageobject.generator.locator;

import com.google.common.base.Preconditions;
import io.pageobject.generator.GeneratorContext;
import io.pageobject.generator.name.NameExtractor;
import org.jsoup.nodes.Element;

import static io.pageobject.generator.locator.LocatorPart.prefixPart;

public abstract class RepeaterLocatorPartGenerator extends AbstractLocatorPartGenerator {

    public RepeaterLocatorPartGenerator(Element element, LocatorSources... locatorSources) {
        super(element, locatorSources);
    }

    @Override
    public LocatorPart onlyPart(GeneratorContext context) {
        return generateLastPart(context, ELEMENT_ARRAY_FINDER, 1);
    }

    @Override
    public LocatorPart firstPart(GeneratorContext context) {
        String name = new NameExtractor().extractValue(element, context);
        Preconditions.checkNotNull(name, "Could not generate name for repeater [%s]", element);
        return prefixPart("this." + name + ".get(rowIndex1)");
    }

    @Override
    public LocatorPart lastPart(GeneratorContext context, int index) {
        return generateLastPart(context, ELEMENT_ARRAY_FINDER_IN_REPEATER, index);
    }

    @Override
    public LocatorPart lastPartForRepeaterElement(GeneratorContext context, int index) {
        return context.isNestedElement() ? middlePart(context, index) : firstPart(context);
    }
}

File: lib/locator/repeater-locator-part-generator.js
'use strict';

const assert = require('assert');
const AbstractLocatorPartGenerator = require('./abstract-locator-part-generator');
const NameExtractor = require('../name/name-extractor');
const prefixPart = require('./locator-part').prefixPart;

class RepeaterLocatorPartGenerator extends AbstractLocatorPartGenerator {

  constructor(element, ...locatorCandidates) {
    super(element, locatorCandidates);
  }

  onlyPart(context) {
    return super.generateLastPart(context, super.elementArrayFinder, 1);
  }

  firstPart(context) {
    let name = new NameExtractor().extractName(context);
    assert(name !== null, `Could not generate name for repeater [${element}]`);
    return prefixPart('this.' + name + '.get(rowIndex1)');
  }

  lastPart(context, index) {
    return super.generateLastPart(context, super.elementArrayFinderInRepeater, index);
  }

  lastPartForRepeaterElement(context, index) {
    return context.isNestedElement() ? this.middlePart(context, index) : this.firstPart(context);
  }

}

module.exports = RepeaterLocatorPartGenerator;